code/modules/reagents/chemistry/holder.dm

```diff
@@ -677,138 +727,325 @@
+
+	if(!reaction_list.len)
+		finish_reacting()
+	else
+		update_total()
 
-	while(reaction_occurred)
+/*
+* This ends a single instance of an ongoing reaction
+*
+* Arguments:
+* * E - the equilibrium that will be ended
+* Returns:
+* * mix_message - the associated mix message of a reaction
+*/
+/datum/reagents/proc/end_reaction(datum/equilibrium/E)
+	//end reaction proc
+	E.reaction.reaction_finish(src, E.reacted_vol)
+	if(E.reaction.mix_sound)
+		playsound(get_turf(my_atom), E.reaction.mix_sound, 80, TRUE)
+	qdel(E)
+	reaction_list -= E
+	//Reaction occured
 	update_total()
-	if(.)
-		SEND_SIGNAL(src, COMSIG_REAGENTS_REACTED, .)
+	SEND_SIGNAL(src, COMSIG_REAGENTS_REACTED, .)
+	return E.reaction.mix_message
+
+/*
+* This stops the holder from processing at the end of a series of reactions (i.e. when all the equilibriums are completed)
+*
+* Also resets reaction variables to be null/empty/FALSE so that it can restart correctly in the future
+*/
+/datum/reagents/proc/finish_reacting()
+	STOP_PROCESSING(SSprocessing, src)
+	is_reacting = FALSE
+	//Cap off values
+	for(var/r in reagent_list)
+		var/datum/reagent/R = r
+		R.volume = round(R.volume, CHEMICAL_VOLUME_ROUNDING)//To prevent runaways.
+	previous_reagent_list = list() //reset it to 0 - because any change will be different now.
+	update_total()
+	handle_reactions() //Should be okay without. Each step checks.
+
+/*
+* Force stops the current holder/reagents datum from reacting
+*
+* Calls end_reaction() for each equlilbrium datum in reaction_list and finish_reacting()
+* Usually only called when a datum is transfered into a NO_REACT container
+*/
+/datum/reagents/proc/force_stop_reacting()
+	var/mix_message = list()
+	for(var/e in reaction_list)
+		var/datum/equilibrium/E = e
+		mix_message += end_reaction(E)
+	if(length(mix_message))
+		my_atom.visible_message("<span class='notice'>[icon2html(my_atom, viewers(DEFAULT_MESSAGE_RANGE, src))] [mix_message.Join()]</span>")
+	finish_reacting()
+
+
+///Checks to see if the reagents has a difference in reagents_list and previous_reagent_list (I.e. if there's a difference between the previous call and the last)
+/datum/reagents/proc/has_changed_state()
+	var/total_matching_reagents = 0
+	for(var/R in previous_reagent_list)
+		if(has_reagent(R))
+			total_matching_reagents++
+	if(total_matching_reagents != reagent_list.len)
+		return TRUE
+	return FALSE
 
+/datum/reagents/proc/update_previous_reaction_list()
+	previous_reagent_list = list()
+	for(var/r in reagent_list)
+		var/datum/reagent/R = r
+		previous_reagent_list += R.type
+
+///Old reaction mechanics, edited to work on one only
+///This is changed from the old - purity of the reagents will affect yield
+/datum/reagents/proc/instant_react(datum/chemical_reaction/selected_reaction)
+	var/list/cached_required_reagents = selected_reaction.required_reagents
+	var/list/cached_results = selected_reaction.results
+	var/datum/cached_my_atom = my_atom
+	var/list/multiplier = INFINITY
```

Comment:
```suggestion
	var/multiplier = INFINITY
```
This is not meant to be typed as a list, since it's a numeric variable, correct?

Reply:
Yes, I'm not sure how I slipped list in there or how it works presently as a list.